src/expect.jl

```diff
@@ -1,181 +1,189 @@
-    alg::Algorithm,
-    ψ::AbstractITensorNetwork,
-    observables::Vector{<:Tuple};
-    (cache!)=nothing,
-    update_cache=isnothing(cache!),
-    cache_update_kwargs = alg == Algorithm("bp") ? default_posdef_bp_update_kwargs(; cache_is_tree = is_tree(ψ)) : default_cache_update_kwargs(alg),
-    cache_construction_kwargs= (;),
-    message_rank = nothing,
-    kwargs...,
+function expect(ψ::Union{TensorNetworkState, BeliefPropagationCache, BoundaryMPSCache}, observable; alg::Union{String, Nothing} = default_alg(ψ), kwargs...)
+    return expect(Algorithm(alg), ψ, observable; kwargs...)
+end
+
+function expect(
+    alg::Union{Algorithm"bp", Algorithm"boundarymps"},
+    ψ::AbstractBeliefPropagationCache,
+    obs::Tuple;
+    bmps_messages_up_to_date = false,
 )
+    op_strings, obs_vs, coeff = collectobservable(obs)
+    iszero(coeff) && return 0
 
-    if alg == Algorithm("boundarymps") && !isnothing(message_rank)
-        cache_construction_kwargs = merge(cache_construction_kwargs, (; message_rank))
-    end
-    if isnothing(cache!)
-        ψIψ = QuadraticFormNetwork(ψ)
-        cache! = Ref(cache(alg, ψIψ; cache_construction_kwargs...))
+    #For boundary MPS, must stay in partition
+    if length(obs_vs) == 1
+        steiner_vs = obs_vs
+    elseif alg == Algorithm("bp")
+        steiner_vs = collect(vertices(steiner_tree(network(ψ), obs_vs)))
+    elseif alg == Algorithm("boundarymps")
+        partitions = unique(partitionvertices(ψ, obs_vs))
+        length(partitions) > 1 && error("Observable support must be within a single partition (row/ column) of the graph for now.")
+        partition = only(partitions)
+        g = partition_graph(ψ, partition)
+        steiner_vs = collect(vertices(steiner_tree(g, obs_vs)))
+
+        if !bmps_messages_up_to_date
+            ψ = update_partition(ψ, partition)
+        end
     end
+    op_string_f = v -> v ∈ obs_vs ? op_strings[findfirst(x->x == v, obs_vs)] : "I"
 
-    if update_cache
-        cache![] = update(cache![]; cache_update_kwargs...)
-    end
+    incoming_ms = incoming_messages(ψ, steiner_vs)
+    ψIψ_tensors = ITensor[norm_factors(network(ψ), steiner_vs); incoming_ms]
+    denom_seq = contraction_sequence(ψIψ_tensors; alg="einexpr", optimizer=Greedy())
+    denom = contract(ψIψ_tensors; sequence=denom_seq)[]
+
+    ψOψ_tensors = ITensor[norm_factors(network(ψ), steiner_vs; op_strings = op_string_f); incoming_ms]
+    numer_seq = contraction_sequence(ψOψ_tensors; alg="einexpr", optimizer=Greedy())
+    numer = contract(ψOψ_tensors; sequence=numer_seq)[]
 
-    return expect(cache![], observables; alg, kwargs...)
+    return coeff * numer/ denom
 end
 
-# Here we turn a single tuple observable into a vector of tuples -- the expected format in ITensorNetworks
-function ITensorNetworks.expect(
-    alg::Algorithm,
-    ψ::AbstractITensorNetwork,
-    observable::Tuple;
+function expect(
+    alg::Algorithm"boundarymps",
+    cache::BoundaryMPSCache,
```

Comment:
What is the reasoning behind always dispatching both on `alg` and the type of `cache`? This can cause confusion about which function is entered and which contraction is actually taking place. I would have thought that `alg` specifies what to do when you pass a state, but otherwise it dispatches onto whatever the type of `cache` is.